# Exercise 06 - Bank Account

|                         |                     |
| -----------------------:| ------------------- |
|   Turn-in directory:    |  ex06               |
|   Files to turn in:     |  the_bank.py        |
|   Forbidden functions:  |  None               |
|   Remarks:              |  n/a                |

It's all about security.  
Have a look at the class named `Account` in the snippet of code below.

## Instructions:

```py
# in the_bank.py
class Account(object):

    ID_COUNT = 1

    def __init__(self, name, **kwargs):
        self.__dict__.update(kwargs)
        
        self.id = self.ID_COUNT
        Account.ID_COUNT += 1
        self.name = name
        if not hasattr(self, 'value'):
            self.value = 0
        
        if self.value < 0:
            raise AttributeError("Attribute value cannot be negative.")
        if not isinstance(self.name, str):
            raise AttributeError("Attribute name must be a str object.")
    
    def transfer(self, amount):
        self.value += amount

```

Now, it is your turn to code a class named `Bank`!  
Its purpose will be to handle the security part of each transfer attempt.

Security means checking if the `Account` is:
* the right object
* not corrupted
* and stores enough money to complete the transfer.

How do we define if a bank account is corrupted? A corrupted bank account has:
* an even number of attributes
* an attribute starting with `b`
* no attribute starting with `zip` or `addr`
* no attribute `name`, `id` and `value`
* `name` not being a string
* `id` not being an `int`
* `value` not being an `int` or a `float`

For the rest of the attributes `addr`, `zip`, etc ... there is no specific check expected, meaning you are not expected to evaluate the validity of the account based on the type of the other attributes except (the conditions listed above are sufficient).

Moreover, verification has to be performed when account objects are added to to Bank instance (`bank.add(Account(...))`).
The verification in `add` only check the type of the new_account and if there is no account among the one already in Bank instance with the same name.

A transaction (`transfer`) is invalid if `amount < 0` or if the amount is larger than the available funds of the sending account. Prior the transfer, the validity of the 2 accounts (`origin` and `dest`) are checked (according to the list of criteria above).
A transfer between the same account (`bank.transfer('Wiliam John', 'William John')`) is valid but there is no fund movement.

`fix_account` recovers a corrupted account if it parameter `name` correspond to the attribute
name of one of the account in `accounts` (attribute of Bank). If name is not a string or does not corresponded
to an account name, the method return `False`.

```py
# in the_bank.py
class Bank(object):
    """The bank"""
    def __init__(self):
        self.accounts = []

    def add(self, new_account):
        """ Add new_account in the Bank
            @new_account:  Account() new account to append
            @return   True if success, False if an error occured
        """
        # test if new_account is an Account() instance and if 
        # it can be appended to the attribute accounts
        
        # ... Your code  ...

        self.accounts.append(new_account)

    def transfer(self, origin, dest, amount):
        """ Perform the fund transfer
            @origin:  str(name) of the first account
            @dest:    str(name) of the destination account
            @amount:  float(amount) amount to transfer
            @return   True if success, False if an error occured
        """
        # ... Your code  ...

    def fix_account(self, name):
        """ fix account associated to name if corrupted
            @name:   str(name) of the account
            @return  True if success, False if an error occured
        """
        # ... Your code  ...
```

Check out the `dir` function.

WARNING: YOU WILL HAVE TO MODIFY THE INSTANCES' ATTRIBUTES IN ORDER TO FIX THEM.

## Examples:
The script `banking_test1.py` is a test which must print `Failed`.
The second script `banking_test2.py` is a test which must print `Failed` and then `Success`.
```bash
>> python banking_test1.py
Failed
# The transaction is not performed has the account of Smith Jane is corrupted (due to the attribute 'bref').

>> python banking_test2.py
Failed
Success
# the account are false due to the abscence of addr attribute, fix_account recover the account,
# thus they become valid.
```